pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,41 +224,234 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
-func (c *MCRouteController) worker() {
+func (c *MCDefaultRouteController) worker() {
 	for c.processNextWorkItem() {
 	}
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
+func (c *MCDefaultRouteController) processNextWorkItem() bool {
 	obj, quit := c.queue.Get()
 	if quit {
 		return false
 	}
 	defer c.queue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
+	key, ok := obj.(string)
+	if !ok {
 		c.queue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.wireGuardConfig != nil {
+		wireGuardWasInitialized := c.wireGuardInitialized
+		if err := c.reconcileWireGuard(key); err != nil {
+			klog.ErrorS(err, "Failed to reconcile Multicluster WireGuard")
+			c.queue.AddRateLimited(key)
+			return true
+		}
+
+		// If the WireGuard was not initialized before and it's initialized after reconcileWireGuard, it's means
+		// that the Gateway has been updated and a new Gateway event will be triggered. We could just skip syncMCFlows.
+		if !wireGuardWasInitialized && c.wireGuardInitialized {
+			klog.InfoS("The Gateway has been updated")
+			return true
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.queue.Forget(key)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", key)
 	}
 	return true
 }
 
-func (c *MCRouteController) syncMCFlows() error {
+// reconcileWireGuard reconcile WireGuard configurations in following way:
+//  1. If the key is a Gateway dummy key, it means that we should check current Node and initialize/remove corresponding
+//     WireGuard interface on the Gateway Node.
+//  2. If the key is not a Gateway dummy key, it means that a new ClusterInfoImport event comes. The controller should
+//     update WireGuard peers on the Gateway Node.
+//
+// Note: MCDefaultRouteController runs only one worker to process Gateway and ClusterInfoImport. So we do not need
+// any synchronization mechanism.
+func (c *MCDefaultRouteController) reconcileWireGuard(key string) error {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get active Gateway")
+		return err
+	}
+
+	isMyselfGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+	isGatewayItem := isGatewayKey(key)
+	if isGatewayItem {
+		// Controller will try to remove potential WireGuard interface on the host
+		// before WireGuard initialization or Gateway is changed or deleted.
+		if c.wireGuardClient != nil && (!isMyselfGateway || !c.wireGuardInitialized) {
+			if err := c.cleanUpWireGuard(); err != nil {
+				return err
+			}
+			c.wireGuardInitialized = false
+		}
+		if !isMyselfGateway {
+			klog.InfoS("Current Node is not Gateway Node, skip WireGuard reconcile")
```

Comment:
You will have this log too, after you delete clean WG when the Gateway changes. Is that what you want? I actually feel we do not need this log. Maybe change to a different log level.

Reply:
Done. Change to log to DEBUG level.